From e6c44d4f7164aed7c10d8b9136eb4841dfdf92a5 Mon Sep 17 00:00:00 2001
From: Rishikesh Donadkar <r-donadkar@ti.com>
Date: Fri, 6 Jun 2025 16:31:56 +0530
Subject: [PATCH 1/2] PENDING: media: ti: j721e-csi2rx: Serialize stream stops

Variable csi->enable_count tracks number of  active streams, it is
read to assert the pixel reset when stopping the last stream.

Currently the value of csi->enable_count is read without
synchronization, and it is decremented after calling disable_streams()
on the downstream element. This creates a race conditions when multiple
streams are stopped simultaneously, which could result in an unnecessary
pixel reset being asserted.

Serialize the read and decrement of csi->enable_count to prevent race
conditions.

Reviewed-by: Devarsh Thakkar <devarsht@ti.com>
Signed-off-by: Rishikesh Donadkar <r-donadkar@ti.com>
---
 .../media/platform/ti/j721e-csi2rx/j721e-csi2rx.c | 15 +++++++--------
 1 file changed, 7 insertions(+), 8 deletions(-)

diff --git a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
index 04db64cd783e..fcbd57830bb4 100644
--- a/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
+++ b/drivers/media/platform/ti/j721e-csi2rx/j721e-csi2rx.c
@@ -1100,6 +1100,7 @@ static void ti_csi2rx_stop_streaming(struct vb2_queue *vq)
 	struct ti_csi2rx_dev *csi = ctx->csi;
 	int ret;
 
+	mutex_lock(&csi->mutex);
 	/* assert pixel reset to prevent stale data */
 	if (csi->enable_count == 1) {
 		writel(0, csi->shim + SHIM_DMACNTX(ctx->idx));
@@ -1116,6 +1117,8 @@ static void ti_csi2rx_stop_streaming(struct vb2_queue *vq)
 	if (ret)
 		dev_err(csi->dev, "Failed to stop subdev stream\n");
 
+	mutex_unlock(&csi->mutex);
+
 	ti_csi2rx_cleanup_buffers(ctx, VB2_BUF_STATE_ERROR);
 	pm_runtime_put(csi->dev);
 }
@@ -1277,19 +1280,15 @@ static int ti_csi2rx_sd_disable_streams(struct v4l2_subdev *sd,
 						       TI_CSI2RX_PAD_SINK,
 						       &streams_mask);
 
-	mutex_lock(&csi->mutex);
-	if (csi->enable_count == 0) {
-		ret = -EINVAL;
-		goto out;
-	}
+	if (csi->enable_count == 0)
+		return -EINVAL;
 
 	ret = v4l2_subdev_disable_streams(csi->source, remote_pad->index,
 					  sink_streams);
 	if (!ret)
 		--csi->enable_count;
-out:
-	mutex_unlock(&csi->mutex);
-	return ret;
+
+	return 0;
 }
 
 static const struct v4l2_subdev_pad_ops ti_csi2rx_subdev_pad_ops = {
-- 
2.50.1

